### DISCUSSION TITLE
Can't migrate Apollo Gateway to Server v4

### DISCUSSION THREAD

**Discussion by tiagodj** [apollographql/apollo-server#7266] [Q&A] [answered]
Hi,

I am having some difficulty migrating my Apollo Gateway to the new Apollo Sever v4.

The issue is that in our code we do a filtering of the Schema before creating the Server, but after creating the Gateway.

The code in v3 is:

```typescript
  const gateway = new ApolloGateway(config);

  const { schema, executor } = await gateway.load();

  const publicSchema = graphqlSFilter(schema, graphqlFilterMap); <---- this filters out APIs from the public schema

  const server = new ApolloServer({
    schema: publicSchema,
    executor,
    ...
});
```

but in the Apollo v4 I was no able to find a configuration where this works.

How do I achieve this in v4?

I am open to a completely different approach.

The main goal is that we have 2 gateways running in parallel: one used by internal apps with all APIs, and one exposed to external public API with limited operations.

**Comment by tiagodj** [comment, marked as answer]
I was able to find a solution! (of course after I asked the question here)

I am converting the SDL to GraphQLSchema and back again before creating the gateway:

```typescript
import { buildSchema } from "graphql";
import { printSchemaWithDirectives } from "@graphql-tools/utils";

  // ...
  const fullSchema = buildSchema(supergraphSdl);
  const filteredSchema = graphqlSFilter(fullSchema);
  const filteredSdl = printSchemaWithDirectives(filteredSchema, {
    assumeValid: true,
    commentDescriptions: false,
  });

  const gateway = new ApolloGateway({
    supergraphSdl: filteredSdl,
    ...
  });
```

**Comment by trevor-scheer** [comment]
Thanks for reporting back with your solution @tiagodj. You're using what we'd consider a custom executor. We do offer a migration path for your use case in [the migration guide](https://www.apollographql.com/docs/apollo-server/migration#executor).

**Reply by nikhiltevaera** [reply]
Hey @trevor-scheer ,  in the current ApolloServer constructor, either you could pass a gateway or a schema, you cannot do both as it was in earlier versions, how to do that??
Do we need apollo gateway for this ?

**Reply by glasser** [reply]
`gateway` is a mechanism for loading a schema, so you cannot provide both.

**Reply by nikhiltevaera** [reply]
Thanks for the reply,

But i am unable to achieve this, lets say i am using this gateway approach 

`new ApolloServer({
  gateway: {
    async load() {
      return { executor };
    },
    onSchemaLoadOrUpdate() {
      return () => {};
    },
    async stop() {},
  },
});`


But the issue with with this is, with my custom executor , i am using schema reference so it says "start server before getDerivedSchema error"

but i cannot call it inside of the executor (because server.start cannot be called twice) , as i am already calling after instantiating the apollo server.

it seems like a dead lock...
Can you suggest any ways out there...

**Reply by glasser** [reply]
I don't think I'll easily be able to help with just pseudocode and not understanding what you are trying to do overall.

**Reply by nikhiltevaera** [reply]
Hey David, 
I have sent you a request on Linkedin ,please let me know if you receive and then on we can continue our conversation.

**Reply by glasser** [reply]
Hi, I don't think private LinkedIn messages are the best way to provide open source support — it's best if any help I give open source users is available to help everyone. Perhaps you can minimize your issue to a small example. If you're an Apollo customer looking for private support, we have a process for that of course.

**Reply by nikhiltevaera** [reply]
Alright then let me provide some example snippets,
`
const schema = createSchema();

const executor: GatewayExecutor = async ({ operationName, queryHash, document, context, request }) => {
    ......
    // the logic here contains caching compiled Query....
    return await compiledQuery.query(schema, context, request.variables || {});
  };
  
  const server = new ApolloServer({
    includeStacktraceInErrorResponses: !isProduction(),
    introspection: true,
    validationRules: [depthLimit(10)], // Restrict queries to a depth of 10.
    plugins: [
      process.env.NODE_ENV === 'production'
        ? ApolloServerPluginLandingPageDisabled()
        : ApolloServerPluginLandingPageLocalDefault({ footer: false }),
    ],
     gateway: {
        async load() {
          return { executor };
        },
        onSchemaLoadOrUpdate() {
          return () => {};
        },
        async stop() {},
      },
    });
    
    await server.start();
    const apolloMiddleware = expressMiddleware(server, {context : ...})
    
    //app is from express
    app.use("/",apolloMiddleware);
`

 
 So during gateway initialization , it throws an error as "Error: You must call start() before getSchemaDerivedData()",
 
 I am building schema and schema using my custom function which returns  schema of type GraphQlSchema.
 
 I want to load schema through gateway...,Need your help on this

**Reply by glasser** [reply]
This is a little hard to follow, but you need actually implement onSchemaLoadOrUpdate — it takes a callback that you pass the schema to when you have it.

**Reply by nikhiltevaera** [reply]
let me make this simpler for you,
Earlier in apollo server 3 - [Migration](https://www.apollographql.com/docs/apollo-server/migration/#executor)

i have my custom executor that i pass similar to the guide shown in here.

, i want also to load schema before the executor initializes itself...

Since i am using schema in my custom_executor , for some logic, so the server throws the issue as -"  You must call start() before getSchemaDerivedData()"
